0007-change-requisites-excludes.md

```diff
@@ -0,0 +1,147 @@
+out of everywhere they're referenced - state files, requirements, etc.
+
+The proposal here is that we change Salt to line up with the user expectations.
+The way this could work is that during the `apply_exclude` process, anything
+that can be excluded will be removed, as if it never existed. Here is a simple
+example state that currently does not work:
+
+    test1:
+      test.nop
+
+    test2:
+      test.nop:
+        - require:
+          - test: test1
+
+    exclude:
+      - id: test1
+
+This fails because when it tries to run test2, the require still exists, but
+refers to a state that is no longer there.
+
+While there are some obvious ways to handle this case, a slightly more subtle
+example was also encountered:
+
+    test1:
+      test.nop
+
+    test2:
+      test.nop:
+        - require_in:
+          - test: test1
+
+    exclude:
+      - id: test2
+
+In this case, it feels quite odd if you're not super familiar with the
+requisites because you're excluding a state, but part of that state has already
+affected another one.
+
+In these cases, the behavior that Salt's users expect is that the exclude will
+take precedence, resulting in these states:
+
+    test2:
+      test.nop
+
+And
+
+    test1:
+      test.nop
+
+In order to preserve backwards compatibility, we [should make this
+configurable][1]. The following command line argument should be added to the
+salt/salt-call cli:
+
+    --exclude-before-requisites
+
+We should also add the same setting to the config files:
+
+    exclude_before_requisites: True
+
+When these flags are set to `True`, excludes should work as proposed. When the
+flags are set to `False`, current behavior should be kept, though we should add
+an explicit message that the exclude would remove a required state.
+
+At first glance, it's *possible* that this code change is as simple as moving
+the `apply_excludes` line in `state.py` a couple of lines earlier before the
+`requisite_in` calls. Further tests indicate there's more considerations.
+
+The test cases are pretty straight forward, though it's possible that it breaks
+some existing tests.
+
+Test cases:
+
+- when the flag is not set, excluding a requisite should produce a helpful
+  error message (`'Warning: excluding {thing} that is required by
+  {other_thing}'`, for example)
+- exclude with no requires still excludes
+- exclude with [each
+  requisite][2]
+  properly excludes when the required state, or the requiring state is excluded
+
+## Alternatives
+[alternatives]: #alternatives
+
+There are a couple of alternative approaches:
+
+1. Do nothing - everything is fine.
+2. Add documentation to either requisites and/or includes that mentions that
+   this is considered an anti-pattern, and provide some patterns to follow for
+   people who think they want to take this approach.
+3. Codify the existing behavior by adding a proper error/exception when exclude
+   excludes something that was required (should still require documentation
```

Comment:
I like the PR, but if we go down this path instead, I'd advocate a warning instead of an error. An error would be a breaking change if someone was relying on current behavior. If an error is really desired, then have one major release use a warning with a deprecation notice first, then make it an error.

Reply:
That's a great point all around. Definitely a warning+deprecation notice, e.g. "This will become an error and <do whatever> in release <whatever>". That would handle the current case, where it's implicit behavior, as well as provide a good path going forward.